src/HttpCacheListener.php

```diff
@@ -273,4 +293,74 @@ public function setVary(Headers $headers)
 
         return $this;
     }
+
+    /**
+     * @param HttpRequest $request
+     * @param HttpResponse $response
+     * @return $this
+     */
+    public function setETag(HttpRequest $request, HttpResponse $response)
+    {
+        $headers = $response->getHeaders();
+
+        if (empty($this->cacheConfig['etag'])) {
+            return $this;
+        }
+
+        // ETag is already set and we should not override, default is to not overwrite.
+        if ($headers->has('Etag')
+            && !empty($this->cacheConfig['etag']['override'])
+            && $this->cacheConfig['etag']['override'] === false) {
+
+            return $this;
+        }
+
+        $generator = $this->getETagGenerator();
+        $headers->addHeader(new Header\Etag($generator->generate($request, $response)));
+
+        return $this;
+    }
+
+    /**
+     * @param HttpRequest $request
+     * @param HttpResponse $response
+     * @return $this
+     */
+    public function setNotModified(HttpRequest $request, HttpResponse $response)
+    {
+        if (!$request->getHeaders()->has('If-None-Match') || !$response->getHeaders()->has('Etag')) {
+            return $this;
+        }
+
+        $requestEtags = $request->getHeaders()->get('If-None-Match')->getFieldValue();
+        $requestEtags = !is_array($requestEtags) ? [$requestEtags] : $requestEtags;
+        $responseEtag = $response->getHeaders()->get('Etag')->getFieldValue();
+
+        if (in_array($responseEtag, $requestEtags) || in_array('*', $requestEtags)) {
+            $response->setStatusCode(304);
+            $response->setContent(null);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Returns an instance of a ETag generator.
+     *
+     * @return ETagGeneratorInterface
+     */
+    protected function getETagGenerator()
```

Comment:
This functionality should be moved into the factory, and the constructor updated to accept an `EtagGeneratorInterface` instance. I'll make that change on merge.